File: src/java/xt/xt/audio/XtAudio.java
package xt.audio;

import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.Pointer;

import xt.audio.Callbacks.XtOnError;
import xt.audio.Enums.XtSample;
import xt.audio.Structs.XtAttributes;
import xt.audio.Structs.XtErrorInfo;
import xt.audio.Structs.XtVersion;
import xt.audio.NativeCallbacks.NativeOnError;
import static xt.audio.Utility.handleAssert;

public final class XtAudio {

    private static XtOnError _onError;
    private static NativeOnError _onNativeError;
    private static void onError(String message) throws Exception { _onError.callback(message); }

    static { Native.register(Utility.LIBRARY); }

    private static native XtVersion.ByValue XtAudioGetVersion();
    private static native void XtAudioSetOnError(NativeOnError onError);
    private static native Pointer XtAudioInit(String id, Pointer window);
    private static native XtErrorInfo.ByValue XtAudioGetErrorInfo(long error);
    private static native XtAttributes.ByValue XtAudioGetSampleAttributes(XtSample sample);

    private XtAudio() {}

    public static XtVersion getVersion() { return handleAssert(XtAudioGetVersion()); }
    public static XtErrorInfo getErrorInfo(long error) { return handleAssert(XtAudioGetErrorInfo(error)); }
    public static XtPlatform init(String id, Pointer window) { return new XtPlatform(handleAssert(XtAudioInit(id, window))); }
    public static XtAttributes getSampleAttributes(XtSample sample) { return handleAssert(XtAudioGetSampleAttributes(sample)); }

    public static void setOnError(XtOnError onError) {
        _onError = onError;
        _onNativeError = (NativeOnError)XtAudio::onError;
        handleAssert(() -> XtAudioSetOnError(_onNativeError));
    }
}

File: src/core/xt/xt/api/XtAudio.cpp
#include <xt/api/XtAudio.h>
#include <xt/shared/Services.hpp>
#include <xt/private/Platform.hpp>

#include <memory>
#include <thread>
#include <cstring>

XtVersion XT_CALL
XtAudioGetVersion(void) 
{ return { 2, 0 }; }
char const* XT_CALL
XtAudioGetLastAssert(void)
{ return XtiGetLastAssert(); }
void XT_CALL
XtAudioSetOnError(XtOnError onError)
{ XtiSetOnError(onError); }
void XT_CALL
XtAudioSetAssertTerminates(XtBool terminates)
{ XtiSetAssertTerminates(terminates); }

XtErrorInfo XT_CALL
XtAudioGetErrorInfo(XtError error) 
{
  XtErrorInfo result;
  XT_ASSERT_API(error != 0);
  auto fault = XtiGetErrorFault(error);
  auto sysid = (error & 0xFFFFFFFF00000000) >> 32ULL;
  auto system = static_cast<XtSystem>(sysid);
  result.fault = fault;
  result.system = system;
  result.service = XtiGetServiceError(system, fault);
  return result;
}

XtAttributes XT_CALL
XtAudioGetSampleAttributes(XtSample sample) 
{
  XtAttributes result;
  XT_ASSERT_API(XtSampleUInt8 <= sample && sample <= XtSampleFloat32);
  result.isSigned = sample != XtSampleUInt8;
  result.isFloat = sample == XtSampleFloat32;
  result.count = sample == XtSampleInt24? 3: 1;
  switch(sample) 
  {
  case XtSampleUInt8: result.size = 1; break;
  case XtSampleInt16: result.size = 2; break;
  case XtSampleInt24: result.size = 3; break;
  case XtSampleInt32: result.size = 4; break;
  case XtSampleFloat32: result.size = 4; break;
  default: XT_ASSERT(false);
  }
  return result;
}

XtPlatform* XT_CALL
XtAudioInit(char const* id, void* window)
{
  XT_ASSERT_API(XtPlatform::instance == nullptr);
  auto result = std::make_unique<XtPlatform>();
  if(!result->Init(window)) return nullptr;
  result->_threadId = std::this_thread::get_id();
  result->_id = id == nullptr || strlen(id) == 0? "XT-Audio": id;
  auto alsa = XtiCreateAlsaService();
  if(alsa) result->_services.emplace_back(std::move(alsa));
  auto jack = XtiCreateJackService();
  if(jack) result->_services.emplace_back(std::move(jack));
  auto asio = XtiCreateAsioService();
  if(asio) result->_services.emplace_back(std::move(asio));
  auto pulse = XtiCreatePulseService();
  if(pulse) result->_services.emplace_back(std::move(pulse));
  auto dsound = XtiCreateDSoundService();
  if(dsound) result->_services.emplace_back(std::move(dsound));
  auto wasapi = XtiCreateWasapiService();
  if(wasapi) result->_services.emplace_back(std::move(wasapi));
  return XtPlatform::instance = result.release();
}